benchmark/benchmark.cpp

```diff
@@ -0,0 +1,377 @@
+          dist_(0, max_index)
+    {
+        rng_.seed(seed);
+    }
+    auto
+    operator()()
+    {
+        return ts_[dist_(rng_)];
+    }
+};
+
+#if WITH_ROCKSDB
+std::map<std::string, std::chrono::duration<double>>
+do_timings_rocks(std::uint64_t num_inserts,
+    std::uint64_t num_fetches,
+    std::uint32_t key_size)
+{
+    std::map<std::string, std::chrono::duration<double>> result;
+    nudb::test::temp_dir td;
+
+    std::unique_ptr<rocksdb::DB> pdb = [path = td.path()]
+    {
+        rocksdb::DB* db = nullptr;
+        rocksdb::Options options;
+        options.create_if_missing = true;
+        auto const status = rocksdb::DB::Open(options, path, &db);
+        if (!status.ok())
+            db = nullptr;
+        return std::unique_ptr<rocksdb::DB>{db};
+    }();
+
+    if (!pdb)
+    {
+        std::cerr << "Failed to open rocks db.\n";
+        return result;
+    }
+
+    auto inserter = [key_size, &db = *pdb](auto const& v)
+    {
+        db.Put(rocksdb::WriteOptions(),
+            rocksdb::Slice(reinterpret_cast<char const*>(v.key), key_size),
+            rocksdb::Slice(reinterpret_cast<char const*>(v.data), v.size));
+    };
+
+    auto fetcher = [&db = *pdb](auto const& v)
+    {
+        std::string value;
+        auto const s = db.Get(rocksdb::ReadOptions(),
+            rocksdb::Slice(
+                reinterpret_cast<char const*>(&v.key), sizeof(v.key)),
+            &value);
+        (void)s;
+        assert(s.ok());
+    };
+
+    nudb::test::test_store ts{key_size, 0, 0};
+    result["insert"] = time_block(
+        num_inserts, gen_key_value{ts, 0}, inserter);
+    result["fetch"] = time_block(
+        num_fetches, rand_existing_key{ts, num_inserts - 1}, fetcher);
+
+    return result;
+}
+#endif
+
+std::map<std::string, std::chrono::duration<double>>
+do_timings(std::uint64_t num_inserts,
+    std::uint64_t num_fetches,
+    std::uint32_t key_size,
+    std::size_t block_size,
+    float load_factor)
+{
+    std::map<std::string, std::chrono::duration<double>> result;
+
+    boost::system::error_code ec;
+
+    try
+    {
+        nudb::test::test_store ts{key_size, block_size, load_factor};
+        ts.create(ec);
+        if (ec)
+            goto fail;
+        ts.open(ec);
+        if (ec)
+            goto fail;
+
+        auto inserter = [&ts, &ec](auto const& v) {
+            ts.db.insert(v.key, v.data, v.size, ec);
+            if (ec)
+                throw boost::system::system_error(ec);
+        };
+
+        auto fetcher = [&ts, &ec](auto const& v) {
+            ts.db.fetch(v.key, [&](void const* data, std::size_t size) {}, ec);
+            if (ec)
+                throw boost::system::system_error(ec);
+        };
+
+        result["insert"] = time_block(
+            num_inserts, gen_key_value{ts, 0}, inserter);
+        result["fetch"] = time_block(
+            num_fetches, rand_existing_key{ts, num_inserts - 1}, fetcher);
+    }
+    catch (boost::system::system_error const& e)
```

Comment:
You should be able to use `nudb::system_error` instead. If not, we should add it (include/nudb/error.hpp)